packages/api/src/auth/signed-data.ts

```diff
@@ -0,0 +1,44 @@
+import { hexToString, stringToHex } from "@polkadot/util";
+import { signatureVerify } from "@polkadot/util-crypto";
+// Import
+import { z } from "zod";
+
+export const SignedDataSchema = z.object({
+  payload: z.string({ description: "in hex" }),
+  signature: z.string({ description: "in hex" }),
+  address: z.string({ description: "in hex" }),
+});
+
+export type SignedInput = z.infer<typeof SignedDataSchema>;
+
+export const verifySignedData = <T extends z.ZodTypeAny>(
+  signedInput: SignedInput,
+  dataSchema: T,
+): { data: z.infer<T>; address: string } => {
+  const { payload, signature, address } = signedInput;
+  const result = signatureVerify(payload, signature, address);
+  if (!result.isValid) {
+    throw new Error("Invalid signature");
+  }
+  const unmarshed = JSON.parse(hexToString(payload));
+  const validated = dataSchema.safeParse(unmarshed);
+  if (!validated.success) {
+    throw new Error(`Invalid payload: ${validated.error.message}`);
+  }
+  return { data: validated.data, address };
+};
+
+export const signData = async <T>(
+  signer: (
+    msgHex: `0x${string}`,
+  ) => Promise<{ signature: `0x${string}`; address: string }>,
+  data: T,
+): Promise<SignedInput> => {
+  const dataHex = stringToHex(JSON.stringify(data));
```

Comment:
I think we may need JSON canonicalization here, but I'm not sure. Probably not.

Reply:
probably not
> Does JSON stringify preserve order?
> Only enumerable own properties are visited. This means Map, Set, etc. will become "{}". You can use the replacer parameter to serialize them to something more useful. Properties are visited using the same algorithm as Object.keys(), which has a well-defined order and is stable across implementations.  **For example, JSON.stringify on the same object will always produce the same string, and JSON.parse(JSON.stringify(obj)) would produce an object with the same key ordering as the original (assuming the object is completely JSON-serializable)**